### PR TITLE
Added resource check functions to check process/system memory

### PR DESCRIPTION
**Pull request by chvink**
Somehow this got missed a couple times. It's needed for a few modules elsewhere and had been sitting uncommitted in my working directory here.

### CODE REVIEW

File: modules/c++/sys/source/ResourceCheck.cpp [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                       double margin)
+{
+    size_t relativeAvailable = systemMemBytes * margin;
+    if (reservedBytes > systemMemBytes)
+    {
+        return 0;
+    }
+
+    size_t absAvailable = systemMemBytes - reservedBytes;
+    return std::min<size_t>(relativeAvailable, absAvailable);
```
Comment (asylvest) [inline modules/c++/sys/source/ResourceCheck.cpp]: Don't need the `<size_t>`

File: modules/c++/sys/source/ResourceCheck.cpp [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    return jobMemBytes < getAvailableMem(systemMemBytes,
+                                         reservedBytes,
+                                         margin);
+}
+
+size_t getAvailableMem(size_t systemMemBytes,
+                       size_t reservedBytes,
+                       double margin)
+{
+    size_t relativeAvailable = systemMemBytes * margin;
```
Comment (asylvest) [inline modules/c++/sys/source/ResourceCheck.cpp]: Can be `const`

File: modules/c++/sys/source/ResourceCheck.cpp [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    if (reservedBytes > systemMemBytes)
+    {
+        return 0;
+    }
+
+    size_t absAvailable = systemMemBytes - reservedBytes;
```
Comment (asylvest) [inline modules/c++/sys/source/ResourceCheck.cpp]: Can be `const`

File: modules/c++/sys/source/ResourceCheck.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
+/* =========================================================================
+ * This file is part of sys-c++ 
+ * =========================================================================
+ * 
+ * (C) Copyright 2004 - 2014, MDA Information Systems LLC
```
Comment (asylvest) [inline modules/c++/sys/source/ResourceCheck.cpp]: Update year

File: modules/c++/sys/include/sys/ResourceCheck.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+ *
+ * \param systemMemBytes the amount of physical RAM on the system
+ * \param reservedBytes the absolute amount of bytes to reserve for system use,
+ *      defaults to 4GB
+ * \param margin the fraction of system RAM that is available for use, defaults
+ *      to 95% (0.95)
+ * \return the number of bytes available 
+ */
+size_t getAvailableMem(size_t systemMemBytes, 
+                       size_t reservedBytes = 4294967296ul,
```
Comment (asylvest) [inline modules/c++/sys/include/sys/ResourceCheck.h]: Will `ul` work ok on Windows where `sizeof(long) == 4` even for 64-bit builds?

File: modules/c++/sys/include/sys/ResourceCheck.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+ *      the system
+ * \param reservedBytes the number of bytes of systemMemBytes to reserve for
+ *      the operating system or other applications, defaults to 4GB
+ * \param margin the fraction of system RAM to assume is available for use,
+ *      defaults to 95% (0.95)
+ * \return true if the job can fit, false otherwise
+ */
+bool canProcessFit(size_t jobMemBytes,
+                   size_t systemMemBytes,
+                   size_t reservedBytes = 4294967296ul,
```
Comment (asylvest) [inline modules/c++/sys/include/sys/ResourceCheck.h]: Ditto

File: modules/c++/sys/include/sys/ResourceCheck.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
+/* =========================================================================
+ * This file is part of sys-c++ 
+ * =========================================================================
+ * 
+ * (C) Copyright 2004 - 2014, MDA Information Systems LLC
```
Comment (asylvest) [inline modules/c++/sys/include/sys/ResourceCheck.h]: Update year

File: modules/c++/sys/source/ResourceCheck.cpp [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public 
+ * License along with this program; If not, 
+ * see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <sys/ResourceCheck.h>
+#include <algorithm>
```
Comment (asylvest) [inline modules/c++/sys/source/ResourceCheck.cpp]: `<algorithm>` goes first

**Comment by chvink** [comment]
Comments should all be addressed

**Pull request merged**